source/common/config/watch_map.cc

```diff
@@ -93,10 +93,36 @@ void WatchMap::onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>
   }
 }
 
+// For responses to on-demand requests, replace the original watch for an alias
+// with one for the resource's name
+void WatchMap::convertAliasWatchToNameWatch(const envoy::api::v2::Resource& resource) {
+  absl::flat_hash_set<Watch*> watches_to_update;
+  for (const auto& alias : resource.aliases()) {
+    const auto interested_watches = watch_interest_.find(alias);
+    if (interested_watches != watch_interest_.end()) {
+      for (const auto& interested_watch : interested_watches->second) {
+        watches_to_update.insert(interested_watch);
+      }
+      break;
+    }
+  }
+  for (const auto& watch : watches_to_update) {
+    updateWatchInterest(watch, {resource.name()});
+  }
+}
+
 void WatchMap::onConfigUpdate(
     const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
     const Protobuf::RepeatedPtrField<std::string>& removed_resources,
     const std::string& system_version_info) {
+  // When an on-demand request is made a Watch is created using an alias, as the resource name isn't
+  // known at that point. When an update containing aliases comes back, we update Watches with
+  // resource names.
+  for (const auto& r : added_resources) {
```

Comment:
Ping @fredlas. Hey, I removed methods I used for on-demand queries, and switched to the "normal" interface. Had to modify watch_map to handle aliases:
 - when an on-demand request is made, a watch is created using <route-config-name/host-header> (aka alias).
 - when a response comes back, the watch is updated with the resource name, which keeps the rest of the implementation as-is. 

Could you take a look when you get a chance?

Reply:
Ok, I think I understand the approach. This should be the right place to do it, and I think you've got a pretty clean idea here.

To summarize my understanding, you are adding a little side logic to `onConfigUpdate` that's like "oh there is some alias stuff, translate that to WatchMap's ordinary non-alias stuff, and then let the ordinary logic handle it".

I think this more specifically works like: all alias-bearing resources also come with actual names, and by the time you are delivering these resources to the callbacks, you have already updated all relevant watches to be watching the actual names. That sounds good.

Reply:
Yep!